.github/workflows/ci.yml

```diff
@@ -0,0 +1,283 @@
+---
+name: "CI"
+on:
+  - "push"
+  - "pull_request"
+
+jobs:
+  black:
+    runs-on: "ubuntu-20.04"
+    env:
+      INVOKE_NAUTOBOT_CHATOPS_LOCAL: "True"
+    steps:
+      - name: "Check out repository code"
+        uses: "actions/checkout@v2"
+      - name: "Set up Python"
+        uses: "actions/setup-python@v2"
+        with:
+          python-version: "3.9"
+      - name: "Install Poetry"
+        run: "pip install poetry"
+      - name: "Cache"
+        uses: "actions/cache@v2"
+        id: "cached-poetry"
+        with:
+          path: "~/.cache/pypoetry/virtualenvs/"
+          key: "{{ runner.os }}-poetry-${{ hashFiles('./poetry.lock') }}"
+      - name: "Install environment"
+        if: "steps.cached-poetry.outputs.cache-hit != 'true'"
```

Comment:
🚀

Reply:
@chadell the `cache` action first checks if the cache with the `key` exists. If not, the output variable `cache-hit` is empty. If the cache with the `key` exists then the cache is copied to the `path`. So the `path` is used only when the `key` exists. In this case the `cache-hit` is set to `true`. I know, complicated... :) I spent quite a lot of time getting the understanding. 

The `key` is generated with the variables as you can see. The key differentiator is the hash that is calculated from the `poetry.lock`. When this file is changed, the key is different and the cache does not exist. So, in that case, the new environment is installed. If the job is executed successfully, the cache is stored as the last step in the job.

We need this step in every job. At first, I thought I can have something like `setup` as the first job, but then I realized that this does not have much sense, as I will still need the `cache` action in every subsequent job.

As far as I understand, YAML anchors are not supported in GitHub actions. But, you can create the `composite` action and then reuse that in your jobs. Can you please check the `gha-custom` branch. I tried to implement this logic. So, I have the action that setup the environment -> either load the environment from the cache or install the environment using `poetry install`.

Reply:
I like the `gha-custom` branch! is there any issue to adopt it?